File: irssicolour.py
#!/usr/bin/env python3

import sys

def main():
	try:
		num = int(sys.argv[1])
	except ValueError:
		print ( "argument must be a number" )
		exit(1)
	except:
		print_palette()
		exit(0)
		#print ( "Usage: {:s} <number>".format(sys.argv[0]) )
	
	if ( num < 0 or num > 255 ):
		print ( "number out of range (0-255)" )
		exit(1)

	print ( cvrt(num) )

def cvrt(num):
	if ( num < 17 ):
		return ( "{:02X}".format(num) )
	else:
		num -=6
	
	i=0
	for section in range( 10, 257, 36 ):
		if ( num < section ):
			break
		i += 1
	
	tens = i * 10
	num -= (tens+((i-1)*26))
	
	# if the remaining number is less than ten just use that
	# if not, use str() to retrieve an ASCII char A-Z - kinda like hexadecimal
	# but 36-based. Above 7x it's diferent again, here it's 0 == A. Bonkers.
	if ( i < 7 ):
		if ( num < 10 ):
			ones = str(num)
		else:
			ones = chr((num-10)+65)
	else:
		ones = chr(num+65)

	return "{}{:s}".format( int(tens/10), ones ) 


def print_palette():
	fg = 15
	linechar = ""
	for col16 in range ( 0, 16 ):
		linechar += "\x1B[38;5;{:d}m\x1B[48;5;{:d}m{:^4s}\x1B[48;5;0m".format(fg, col16, cvrt(col16)) 
		fg = 0
	
	print ( linechar + "\n" )	
	
	for hblock in [16, 124]:
		for line in range(hblock, hblock+6*6, 6):
			linechar = ""
			if ( line == 16 or line == 124 ):
				fg = 15
			elif ( line == 34 or line == 142 ):
				fg = 0
			for block in range ( 0, 18, 6 ):
				blockstart = block * 6 + line
				for num in range ( blockstart, blockstart+6 ):
					linechar += "\x1B[38;5;{:d}m\x1B[48;5;{:d}m{:^4s}\x1B[48;5;0m".format(fg, num, cvrt(num)) 
		
				linechar += "  "
			
			print ( linechar )
		print ( "" )
	
	
	fg = 15
	linechar = ""
	for greyscale in range ( 232, 256 ):
		linechar += "\x1B[38;5;{:d}m\x1B[48;5;{:d}m{:^4s}\x1B[48;5;0m".format(fg, greyscale, cvrt(greyscale)) 
		if ( greyscale == 243 ):
			linechar += "\n"
			fg = 0
	
	print ( linechar )

main()

File: irssicolour.go
package main

import "fmt"
import "flag"
import "strconv"
import "os"

type ConvFunc func(int) string

var conv ConvFunc

func main() {
	irssi := flag.Bool("i", true, "Print irssi %xNN palette")
	palette256 := flag.Bool("n", false, "Print colourN palette")
	flag.Parse()

	if flag.NArg() > 0 {
		num, _ := strconv.Atoi(flag.Args()[0])
		if -1 < num && num < 256 {
			fmt.Println(convert(num))
		} else {
			fmt.Println("Argument ", num, " out of range")
			os.Exit(1)
		}
	} else {
		if *palette256 == true {
			conv = pass
		} else if *irssi == true {
			conv = convert
		}
		palette()
	}
}

func palette() {
	fg := 15
	for col := 0; col < 17; col++ {
		fmt.Print(printField(fg, col))
		fg = 0
	}
	fmt.Println("\x1B[93;65m\n")

	for _, hblock := range []int{16, 124} {
		for line := hblock; line < hblock+6*6; line += 6 {
			linechar := ""
			switch line {
			case 16, 124:
				fg = 15
			case 34, 142:
				fg = 0
			}
			for block := 0; block < 18; block += 6 {
				blockstart := block*6 + line
				for num := blockstart; num < blockstart+6; num++ {
					linechar += printField(fg, num)
				}
				linechar += "  "
			}
			fmt.Println(linechar)
		}
		fmt.Println()
	}
	fg = 15
	linechar := ""
	for greyscale := 232; greyscale < 256; greyscale++ {
		linechar += printField(fg, greyscale)
		if greyscale == 243 {
			linechar += "\n"
			fg = 0
		}
	}
	fmt.Print(linechar, "\n")
}

func printField(fg int, num int) string {
	return fmt.Sprintf("\x1B[38;5;%dm\x1B[48;5;%dm %s\x1B[48;5;0m", fg, num, conv(num))

}

func pass(num int) string {
	return fmt.Sprintf("%3d", num)
}

func convert(num int) string {
	if num < 17 {
		return fmt.Sprintf("%02X ", num)
	}
	num -= 6

	i := 1
	for section := 46; section < 257; section += 36 {
		if num < section {
			break
		}
		i++
	}

	tens := i * 10
	num -= (tens + ((i - 1) * 26))

	ones := ""
	if i < 7 {
		if num < 10 {
			ones = strconv.Itoa(num)
		} else {
			ones = string((num - 10) + 65)
		}
	} else {
		ones = string(num + 65)
	}
	return fmt.Sprintf("%d%s ", tens/10, ones)
}
